atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/async/query/GetQuerySpec.java

```diff
@@ -38,67 +36,74 @@
+     * transferred data.
      */
     private static final String QUERY_FORMAT = "SELECT value, column2 FROM \"%s\".\"%s\" "
             + "WHERE key = :row AND column1 = :column AND column2 > :timestamp "
-            + "ORDER BY column1, column2 ASC "
             + "LIMIT 1;";
 
-    @org.immutables.value.Value.Auxiliary
-    public abstract ByteBuffer row();
-
-    @org.immutables.value.Value.Auxiliary
-    public abstract ByteBuffer column();
+    @Override
+    public String formatQueryString() {
+        return String.format(
+                QUERY_FORMAT,
+                cqlQueryContext().keyspace(),
+                AbstractKeyValueService.internalTableName(cqlQueryContext().tableReference()));
+    }
 
-    @org.immutables.value.Value.Auxiliary
-    public abstract long humanReadableTimestamp();
+    @Override
+    public Statement makeExecutableStatement(PreparedStatement preparedStatement) {
+        return preparedStatement.bind()
+                .setBytes("row", getQueryParameters().row())
+                .setBytes("column", getQueryParameters().column())
+                .setLong("timestamp", getQueryParameters().queryTimestamp());
+    }
 
     @Override
     public ConsistencyLevel queryConsistency() {
         return ConsistencyLevel.LOCAL_QUORUM;
     }
 
     @Override
-    public final Supplier<RowStreamAccumulator<Optional<Value>>> rowStreamAccumulatorFactory() {
+    public Supplier<RowStreamAccumulator<Optional<Value>>> rowStreamAccumulatorFactory() {
         return GetQueryAccumulator::new;
     }
 
-    @Override
-    public String formatQueryString() {
-        return String.format(QUERY_FORMAT, keySpace(), AbstractKeyValueService.internalTableName(tableReference()));
-    }
+    @org.immutables.value.Value.Auxiliary
+    public abstract GetQueryParameters getQueryParameters();
 
-    @Override
-    public Statement makeExecutableStatement(PreparedStatement preparedStatement) {
-        return preparedStatement.bind()
-                .setBytes("row", row())
-                .setBytes("column", column())
-                .setLong("timestamp", queryTimestamp());
-    }
+    @org.immutables.value.Value.Immutable
+    public interface GetQueryParameters {
+        ByteBuffer row();
+
+        ByteBuffer column();
 
-    private long queryTimestamp() {
-        return ~humanReadableTimestamp();
+        long humanReadableTimestamp();
+
+        default long queryTimestamp() {
+            return ~humanReadableTimestamp();
+        }
     }
 
-    @ThreadSafe
-    public static class GetQueryAccumulator implements RowStreamAccumulator<Optional<Value>> {
+    private static class GetQueryAccumulator implements RowStreamAccumulator<Optional<Value>> {
 
-        private Optional<Value> resultValue = Optional.empty();
+        private volatile Value resultValue = null;
+        private volatile boolean assigned = false;
```

Comment:
Non-actionable: Depending on usage this may not need to be volatile (if it's guaranteed that calls to `accumulateRowStream` aren't concurrent, even if it may accidentally be called more than once), which could lead to probably a very very minor perf improvement. Don't do this unless we see that it somehow becomes a problem